components/bootloaders/STM/stm32f103/include/Utilities.h

```diff
@@ -0,0 +1,151 @@
+#pragma once
+#include "HW.h"
+#include "Types.h"
+
+#define UNUSED(x)                ((void)x)
+#define SATURATE(x, min, max)    (x > max) ? (max) : ((x < min) ? (min) : (x))
+// Count number of items in array
+#define COUNTOF(x)           ((uint16_t)(sizeof(x) / sizeof(x[0]))) // return size of an array as uint16
+
+// what is this?
```

Comment:
Reading material:
https://lxr.linux.no/#linux+v2.6.31/include/linux/compiler-gcc.h#L12
https://en.wikipedia.org/wiki/Memory_ordering
Discussion: https://stackoverflow.com/questions/14950614/working-of-asm-volatile-memory

It seems it stops memory access ordering optimization at compile time. Helps make a stricter memory access flow, however does not prevent against CPU instruction optimization

Reply:
nice, very interesting